File: rest-fights/src/test/java/io/quarkus/sample/superheroes/fight/mapping/LocationMapperTests.java
package io.quarkus.sample.superheroes.fight.mapping;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;
import org.mapstruct.factory.Mappers;

import io.quarkus.sample.superheroes.fight.FightLocation;
import io.quarkus.sample.superheroes.location.grpc.LocationType;

class LocationMapperTests {
  private static final String DEFAULT_LOCATION_NAME = "Gotham City";
  private static final String DEFAULT_LOCATION_DESCRIPTION = "An American city rife with corruption and crime, the home of its iconic protector Batman.";
  private static final String DEFAULT_LOCATION_PICTURE = "gotham_city.png";
  private static final io.quarkus.sample.superheroes.location.grpc.Location GRPC_LOCATION = io.quarkus.sample.superheroes.location.grpc.Location.newBuilder()
    .setName(DEFAULT_LOCATION_NAME)
    .setDescription(DEFAULT_LOCATION_DESCRIPTION)
    .setPicture(DEFAULT_LOCATION_PICTURE)
    .setType(LocationType.PLANET)
    .build();

  LocationMapper mapper = Mappers.getMapper(LocationMapper.class);

  @Test
  public void mappingWorks() {
    assertThat(this.mapper.fromGrpc(GRPC_LOCATION))
      .isNotNull()
      .usingRecursiveComparison()
      .isEqualTo(new FightLocation(DEFAULT_LOCATION_NAME, DEFAULT_LOCATION_DESCRIPTION, DEFAULT_LOCATION_PICTURE));
  }
}

File: grpc-locations/src/test/kotlin/io/quarkus/sample/superheroes/location/mapping/LocationMapperTests.kt
package io.quarkus.sample.superheroes.location.mapping

import io.quarkus.sample.superheroes.location.Location
import io.quarkus.sample.superheroes.location.LocationType
import io.quarkus.sample.superheroes.location.grpc.location
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class LocationMapperTests {
	companion object {
		private const val DEFAULT_ID = 1L
		private const val DEFAULT_NAME = "Gotham City"
		private const val DEFAULT_DESCRIPTION = "Where Batman lives"
		private const val DEFAULT_PICTURE = "gotham_city.png"
		private val DEFAULT_TYPE = LocationType.CITY

		private fun createDefaultLocation() : Location {
			val location = Location()
			location.id = DEFAULT_ID
			location.name = DEFAULT_NAME
			location.description = DEFAULT_DESCRIPTION
			location.picture = DEFAULT_PICTURE
			location.type = DEFAULT_TYPE

			return location
		}

		private fun createDefaultGrpcLocation() = location {
      name = DEFAULT_NAME
      description = DEFAULT_DESCRIPTION
      picture = DEFAULT_PICTURE
      type = io.quarkus.sample.superheroes.location.grpc.LocationType.CITY
    }
	}

	@Test
	fun `mapper works correctly for null (model to gRPC)`() {
		assertThat(LocationMapper.toGrpcLocationMaybeNull(createDefaultLocation()))
			.isNotNull
			.extracting(
				"name",
				"description",
				"picture",
				"type"
			)
			.containsExactly(
				DEFAULT_NAME,
				DEFAULT_DESCRIPTION,
				DEFAULT_PICTURE,
				io.quarkus.sample.superheroes.location.grpc.LocationType.CITY
			)

		assertThat(LocationMapper.toGrpcLocationMaybeNull(null))
			.isNull()
	}

	@Test
	fun `mapper works correctly for non-null(model to gRPC)`() {
		assertThat(LocationMapper.toGrpcLocation(createDefaultLocation()))
			.isNotNull
			.extracting(
				"name",
				"description",
				"picture",
				"type"
			)
			.containsExactly(
				DEFAULT_NAME,
				DEFAULT_DESCRIPTION,
				DEFAULT_PICTURE,
				io.quarkus.sample.superheroes.location.grpc.LocationType.CITY
			)
	}

	@Test
	fun `mapper works correctly (gRPC to model)`() {
		assertThat(LocationMapper.fromGrpcLocation(createDefaultGrpcLocation()))
			.isNotNull
			.extracting(
				"name",
				"description",
				"picture",
				"type"
			)
			.containsExactly(
				DEFAULT_NAME,
				DEFAULT_DESCRIPTION,
				DEFAULT_PICTURE,
				DEFAULT_TYPE
			)
	}
}
